### ISSUE TITLE
Sitemaps API

### ISSUE THREAD

**Issue by mcshanea** [GoogleChromeLabs/wp-sitemaps#25]
An XML Sitemaps API will be introduced as part of the integration allowing extensibility. At a high level, below is a list of the ways the XML Sitemaps may be manipulated via the API:

- [x] Add extra sitemaps and sitemap entries
- [x] Add extra attributes to sitemap entries
- [x] Provide a custom XML Stylesheet
- [x] Exclude a specific post type from the sitemap
- [x] Exclude a specific post from the sitemap
- [x] Exclude a specific taxonomy from the sitemap
- [x] Exclude a specific term from the sitemap
- [x] Exclude a specific author from the sitemap
- [x] Exclude a specific authors with a specific role from the sitemap

**Acceptance criteria**

For each of these, we want to ensure we have an implementation that is documented and includes unit test coverage.

**Comment by kirstyburgoine** [comment]
MAking a note so I don't forget later. One filter to potentially add is the ability to change the url of the sitemap index. For example, in Pascal's POC, this is added as part of the method to build the sitemap index URL for use in robots.txt:

```
	/**
	 * Filters the XML sitemap URL.
	 *
	 * @since 1.0.0
	 *
	 * @param string $url            The XML sitemap URL.
	 * @param string $object_type    Object type, e.g. post or term. '1' for the sitemap index.
	 * @param string $object_subtype Object sub type, e.g. page or attachment.
	 * @param int    $index          The index of the requested sitemap.
	 */
	return apply_filters( 'sitemaps_sitemap_url', $url, $object_type, $object_subtype, $index );
```

lines 58 - 68 in `sitemap-poc/includes/functions.php`

**Comment by joemcgill** [comment]
First step here is to review what is already in place and generate tickets for anything that isn't already covered.

**Comment by joemcgill** [comment]
I've gone through the list and most of these are already in place, but just need to have the filters documented and unit test coverage added.


## To be implemented

- [x] Add extra sitemaps and sitemap entries
- [x] Add extra attributes to sitemap entries

## Already implemented

The following extensibility options already exist but need documentation and unit test coverage.

**Provide a custom XML Stylesheet**

Document/add coverage for the following filters:

- `core_sitemaps_stylesheet_content`
- `core_sitemaps_index_stylesheet_content`
- `core_sitemaps_stylesheet_css`

**Exclude a specific post type from the sitemap**

- Document and Unit test: `core_sitemaps_post_types` filter.

**Exclude a specific post from the sitemap**

- Document and Unit test: `core_sitemaps_post_url_list` filter.

**Exclude a specific taxonomy from the sitemap**

- Document and Unit test: `core_sitemaps_taxonomies` filter.

**Exclude a specific term from the sitemap**

- Document and Unit test: `core_sitemaps_taxonomies_url_list` filter.

**Exclude a specific author from the sitemap**

- Document and Unit test: `core_sitemaps_taxonomies` filter.

**Exclude a specific authors with a specific role from the sitemap**

- Document and Unit test: `core_sitemaps_users_url_list` filter.

**Comment by joemcgill** [comment]
I've split the remaining work for this ticket into three new tasks (#86, #87, #88) so they can be estimated and worked on in parallel.

For the purposes of the Spike ticket in Sprint 3, this is now done, but we'll leave this open as a tracking ticket until all three sub-tasks are completed.

**Comment by kirstyburgoine** [comment]
Just adding a note, that for **Provide a custom XML Stylesheet** there are also the following filters:
- `core_sitemaps_stylesheet_url`
- `core_sitemaps_stylesheet_index_url`

These can be found in `Core_Sitemaps_Renderer`. However, the intention of these filters were to allow developers to replace the URL of the xsl file being directly linked to in the xml. As the implementation has changed and we now use redirects and the content of the xsl file can be filtered out from within the `Core_Sitemaps_Stylesheet` class these should probably be removed.

**Comment by RuxandraGradina** [comment]
@joemcgill please provide the input on the progress we've made so far and what's the outstanding work we need to do for V1 completion.

**Comment by joemcgill** [comment]
@RuxandraGradina all of that info is documented above in https://github.com/GoogleChromeLabs/wp-sitemaps/issues/25#issuecomment-557305334 and https://github.com/GoogleChromeLabs/wp-sitemaps/issues/25#issuecomment-557317451.

**Comment by joemcgill** [comment]
Each part of the API has been implemented with the completion of #87 and #88. All base functionality has unit test coverage as of #81.

Additional test coverage would be useful for these specific use cases and could be handled in #86. User documentation could also be handled in that ticket or split out into a separate ticket:

- Provide a custom XML Stylesheet
- Exclude a specific post type from the sitemap
- Exclude a specific post from the sitemap
- Exclude a specific taxonomy from the sitemap
- Exclude a specific term from the sitemap
- Exclude a specific author from the sitemap
- Exclude a specific authors with a specific role from the sitemap

**Comment by ThierryA** [comment]
@joemcgill can this be marked as done?

**Comment by svandragt** [comment]
This can be closed as the work originally agreed has been completed.
